File: dtb-api/store-service/src/main/java/app/roaim/dtbazar/storeservice/domain/Donation.java
package app.roaim.dtbazar.storeservice.domain;

import app.roaim.dtbazar.storeservice.model.Currency;
import lombok.Data;
import org.springframework.data.annotation.CreatedDate;
import org.springframework.data.annotation.Id;
import org.springframework.data.annotation.PersistenceConstructor;
import org.springframework.data.mongodb.core.index.Indexed;
import org.springframework.data.mongodb.core.mapping.Document;

import java.time.LocalDateTime;
// TODO move to a separate micro service
@Data
@Document(collection = "donation")
public class Donation {

    @Id
    private String id;
    @Indexed
    private String storeId;
    private String storeName;
    @Indexed
    private String donorId;
    private String donorName;
    @Indexed
    private String storeFoodId;
    private String foodName;
    private double amount;
    private Currency currency = Currency.BDT;
    @CreatedDate
    private LocalDateTime createdAt = LocalDateTime.now();

    @PersistenceConstructor
    public Donation(String storeId, String storeName, String donorId, String donorName, String storeFoodId, String foodName, double amount, Currency currency) {
        super();
        update(storeId, storeName, donorId, donorName, storeFoodId, foodName, amount, currency);
    }

    public void update(Donation donation) {
        update(donation.getStoreId(), donation.getStoreName(), donation.getDonorId(), donation.getDonorName(),
                donation.getStoreFoodId(), donation.getFoodName(), donation.getAmount(), donation.getCurrency());
    }

    public void update(String storeId, String storeName, String giverId, String giverName, String foodId, String foodName, double amount, Currency currency) {
        this.storeId = storeId;
        this.storeName = storeName;
        this.donorId = giverId;
        this.donorName = giverName;
        this.storeFoodId = foodId;
        this.foodName = foodName;
        this.amount = amount;
        this.currency = currency;
    }
}

File: dtb-android/app/src/main/java/app/roaim/dtbazar/model/Donation.kt
package app.roaim.dtbazar.model

import androidx.room.Entity
import androidx.room.Ignore
import com.google.gson.annotations.SerializedName

@Entity(tableName = "donation", primaryKeys = ["id"])
data class Donation(

    @field:SerializedName("createdAt")
    val createdAt: String? = null,

    @field:SerializedName("donorName")
    val donorName: String? = null,

    @field:SerializedName("amount")
    val amount: Double? = null,

    @field:SerializedName("foodName")
    val foodName: String? = null,

    @field:SerializedName("storeFoodId")
    val storeFoodId: String? = null,

    @field:SerializedName("currency")
    val currency: String? = null,

    @field:SerializedName("donorId")
    val donorId: String? = null,

    @field:SerializedName("storeName")
    val storeName: String? = null,

    @field:SerializedName("id")
    val id: String,

    @field:SerializedName("storeId")
    val storeId: String? = null,

    @field:SerializedName("enabled")
    val enabled: Boolean? = null
) : ListItem {
    @Ignore
    override fun getItemId(): String = id
}
